packages/ui/src/components/Table/components/RowSelector.tsx

```diff
@@ -0,0 +1,109 @@
+import { Checkbox, Radio } from '@mui/material';
+import { ChangeEvent, ReactNode, useMemo } from 'react';
+
+import { CheckboxRendererArgs, GenericRowStructure, Row } from '../types';
+
+interface HeadRowSelectorProps<T extends GenericRowStructure> {
+    selectedRowsCount: number;
+    rows: Row<T>[];
+    handleAllRowsSelection: (e: ChangeEvent<HTMLInputElement>, checked: boolean) => void;
+    renderCheckbox?: (args: CheckboxRendererArgs) => ReactNode;
+}
+interface RowSelectorProps<T extends GenericRowStructure> {
+    selectionType: 'single' | 'multiple';
+    row: Row<T>;
+    selectedRowIdsState: Record<string, boolean>;
+    handleRowSelection: (rowId: string, selected: boolean) => void;
+    renderCheckbox?: (args: CheckboxRendererArgs) => ReactNode;
+}
+
+export const HeadRowSelector = <T extends GenericRowStructure>({
+    renderCheckbox,
+    selectedRowsCount,
+    rows,
+    handleAllRowsSelection,
+}: HeadRowSelectorProps<T>) => {
+    const checkboxProps = useMemo(
+        () => ({
+            checked: selectedRowsCount === rows.length,
+            indeterminate: selectedRowsCount > 0 && selectedRowsCount < rows.length,
+            onChange: handleAllRowsSelection,
+        }),
+        [selectedRowsCount, rows.length, handleAllRowsSelection]
+    );
+    const CustomComponent = useMemo(() => {
+        if (!renderCheckbox) return null;
+        return renderCheckbox({ ...checkboxProps });
+    }, [renderCheckbox, checkboxProps]);
+
+    if (!CustomComponent) return <Checkbox {...checkboxProps} inputProps={{ 'aria-label': 'select all rows' }} />;
+    return CustomComponent;
```

Comment:
this is a bit shorter and doesn't use memo since we don't really need it:

```
    if (renderCheckbox) { 
      return renderCheckbox({ ...checkboxProps });
    }
    return <Checkbox {...checkboxProps} inputProps={{ 'aria-label': 'select all rows' }} />;
```